src/library_openal.js

```diff
@@ -1,1622 +1,4591 @@
+          listener._up[0], listener._up[1], listener._up[2]);
+      }
 
-    var ctx;
-    try {
-      ctx = new AudioContext();
-    } catch (e) {
-      try {
-        ctx = new webkitAudioContext();
-      } catch (e) {}
-    }
-
-    if (ctx) {
-      // Old Web Audio API (e.g. Safari 6.0.5) had an inconsistently named createGainNode function.
-      if (typeof(ctx.createGain) === 'undefined') ctx.createGain = ctx.createGainNode;
-
-      var gain = ctx.createGain();
-      gain.connect(ctx.destination);
-      // Extend the Web Audio API AudioListener object with a few tracking values of our own.
-      ctx.listener._position = [0, 0, 0];
-      ctx.listener._velocity = [0, 0, 0];
-      ctx.listener._orientation = [0, 0, 0, 0, 0, 0];
-      var context = {
-        ctx: ctx,
-        err: 0,
-        src: {},
-        buf: [],
-        interval: setInterval(function() { AL.updateSources(context); }, AL.QUEUE_INTERVAL),
-        gain: gain
-      };
-      AL.contexts.push(context);
-      return AL.contexts.length;
-    } else {
-      return 0;
-    }
-  },
+      // Update sources that are relative to the listener
+      for (var i in ctx.sources) {
+        AL.updateSourceSpace(ctx.sources[i]);
+      }
+    },
 
-  alGetError: function() {
-    if (!AL.currentContext) {
-      return 0xA004 /* AL_INVALID_OPERATION */;
-    } else {
-      // Reset error on get.
-      var err = AL.currentContext.err;
-      AL.currentContext.err = 0 /* AL_NO_ERROR */;
-      return err;
-    }
-  },
+    updateSourceSpace: function(src) {
+      if (!src.panner) {
+        return;
+      }
+      var panner = src.panner;
 
-  alcGetError: function(device) {
-    var err = AL.alcErr;
-    AL.alcErr = 0;
-    return err;
-  },
+      var posX = src.position[0];
+      var posY = src.position[1];
+      var posZ = src.position[2];
+      var dirX = src.direction[0];
+      var dirY = src.direction[1];
+      var dirZ = src.direction[2];
 
-  alcGetIntegerv: function(device, param, size, data) {
-    if (size == 0 || !data) {
-      AL.currentContext.err = 0xA003 /* AL_INVALID_VALUE */;
-      return;
-    }
+      var listener = src.context.audioCtx.listener;
+      var lPosX = listener._position[0];
+      var lPosY = listener._position[1];
+      var lPosZ = listener._position[2];
 
-    switch(param) {
-    case 0x1000 /* ALC_MAJOR_VERSION */:
-      {{{ makeSetValue('data', '0', '1', 'i32') }}};
-      break;
-    case 0x1001 /* ALC_MINOR_VERSION */:
-      {{{ makeSetValue('data', '0', '1', 'i32') }}};
-      break;
-    case 0x1002 /* ALC_ATTRIBUTES_SIZE */:
-      if (!device) {
-        AL.alcErr = 0xA001 /* ALC_INVALID_DEVICE */;
-        return 0;
+      // WebAudio expects world space coordinates, so if the source is listener-relative
+      // we must transform the coordinates from listener space into world space.
+      if (src.relative) {
+        // Negate the listener direction since forward is -Z.
```

Comment:
This comment is somewhat ambiguous. Here it would be good to document what the different coordinate spaces are for Web Audio and OpenAL, to illustrate what the needed coordinate transform that is performed here is.